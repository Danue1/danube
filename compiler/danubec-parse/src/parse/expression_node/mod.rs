mod literal_expression_node;

use crate::State;

impl crate::Context {
    pub fn expression_node(&mut self) -> State {
        one_of!(self.expression_literal_node())
    }
}
