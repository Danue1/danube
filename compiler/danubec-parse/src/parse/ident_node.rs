use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn ident_node(&mut self) -> State {
        guard!(self, IDENT_KEYWORD, IDENT_NODE);
        self.finish_node();
        State::Stop
    }
}
