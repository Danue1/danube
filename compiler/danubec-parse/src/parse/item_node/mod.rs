mod enum_item_node;
mod struct_item_node;

impl crate::Context {
    pub fn item_node(&mut self) -> crate::State {
        one_of!(self.struct_item_node(), self.enum_item_node())
    }
}
