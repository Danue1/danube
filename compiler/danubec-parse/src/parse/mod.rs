mod ident_node;
mod item_node;
mod path_node;
mod type_node;

use crate::context::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::context::Context {
    #[inline]
    pub fn ast(&mut self) {
        self.start_node(SyntaxKind::AST_NODE);
        self.skip_whitespace();
        while !self.is_empty() {
            if self.item_node() == State::Continue {
                self.unexpected_token();
            }
            self.skip_whitespace();
        }
        self.finish_node();
    }

    fn unexpected_token(&mut self) {
        self.start_node(SyntaxKind::ERROR);
        self.bump();
        self.finish_node();
    }
}
