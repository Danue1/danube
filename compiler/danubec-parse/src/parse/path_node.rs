use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn path_node(&mut self) -> State {
        guard!(self, IDENT_KEYWORD, PATH_NODE, ident_node);

        while !self.is_eof() {
            if expect!(self, SyntaxKind::COLON) {
                if expect!(self, SyntaxKind::COLON) {
                    self.skip_whitespace();
                    self.ident_node();
                } else {
                    self.unexpected_token();
                }
            } else {
                break;
            }
        }

        self.finish_node();
        State::Stop
    }
}
