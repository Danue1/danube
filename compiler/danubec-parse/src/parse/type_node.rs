use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn type_node(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.unnamed_type_kind_node() == State::Continue
            && self.path_type_kind_node() == State::Continue
        {
            State::Continue
        } else {
            self.start_node_at(checkpoint, SyntaxKind::TYPE_NODE);
            self.finish_node();
            State::Stop
        }
    }

    // ()
    // (type1)
    // (type1, type2)
    pub fn unnamed_type_kind_node(&mut self) -> State {
        guard!(self, LEFT_PAREN, UNNAMED_TYPE_KIND_NODE);
        self.skip_whitespace();

        while !self.is_eof() {
            if expect!(self, SyntaxKind::RIGHT_PAREN) {
                break;
            }

            let checkpoint = self.checkpoint();
            if self.type_node() == State::Continue {
                self.unexpected_token();
            } else {
                self.start_node_at(checkpoint, SyntaxKind::UNNAMED_TYPE_KIND_NODE);
                self.finish_node();
            }
            self.skip_whitespace();
            if expect!(self, SyntaxKind::COMMA) {
                self.skip_whitespace();
            }
        }

        self.finish_node();
        State::Stop
    }

    // Foo
    // Foo::Bar
    pub fn path_type_kind_node(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.path_node() == State::Continue {
            return State::Continue;
        }

        self.start_node_at(checkpoint, SyntaxKind::PATH_TYPE_KIND_NODE);
        self.finish_node();
        State::Stop
    }
}
