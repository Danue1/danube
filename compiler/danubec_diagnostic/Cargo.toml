[package]
name = "danubec_diagnostic"
version = "0.1.0"
edition = "2024"

[dependencies]
