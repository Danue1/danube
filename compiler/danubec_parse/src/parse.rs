use crate::{
    event::Event,
    grammar::{Parser, krate},
    lower::lower_krate,
};
use danubec_ast::Krate;
use danubec_diagnostic::Diagnostic;
use danubec_lex::lex;
use danubec_syntax::{AstNode, GreenNodeBuilder, SyntaxKind, SyntaxNode};
use std::path::PathBuf;

pub fn parse_crate_main(path: &PathBuf) -> Result<Krate, Diagnostic> {
    parse_krate(path.join("src/main.dnb"))
}

pub fn parse_crate_lib(path: &PathBuf) -> Result<Krate, Diagnostic> {
    parse_krate(path.join("src/lib.dnb"))
}

pub fn parse_krate(parent: PathBuf) -> Result<Krate, Diagnostic> {
    let mut diagnostic = Diagnostic::new();
    let source = match read(&parent, &mut diagnostic) {
        Ok(source) => source,
        Err(()) => return Err(diagnostic),
    };
    let (node, mut diagnostic) = parse(&source, diagnostic);
    let krate = match danubec_syntax::Krate::cast(node) {
        Some(node) => lower_krate(node, &mut diagnostic).map_err(|_| diagnostic)?,
        None => {
            diagnostic.report(miette!("ICE: Expected krate node"));
            return Err(diagnostic);
        }
    };

    Ok(krate)
}

fn read(path: &PathBuf, diagnostic: &mut Diagnostic) -> Result<String, ()> {
    match std::fs::read_to_string(path) {
        Ok(source) => Ok(source),
        Err(error) => {
            diagnostic.report(miette!("Failed to read file {}: {}", path.display(), error));

            Err(())
        }
    }
}

pub(crate) fn parse(source: &str, diagnostic: Diagnostic) -> (SyntaxNode, Diagnostic) {
    let tokens = lex(source);
    let (events, diagnostic) = {
        let tokens: Vec<_> = tokens.iter().map(|&(kind, _)| kind).collect();
        let mut parser = Parser::new(&tokens, diagnostic);
        krate(&mut parser);
        parser.finish()
    };
    let node = build(&tokens, events);

    (node, diagnostic)
}

fn build(mut tokens: &[(SyntaxKind, &str)], events: Vec<Event>) -> SyntaxNode {
    let mut builder = GreenNodeBuilder::new();

    macro_rules! advance {
        () => {
            if let Some(&(kind, text)) = tokens.get(0) {
                builder.token(kind.into(), text);
                tokens = &tokens[1..];
            }
        };
    }

    macro_rules! trivia {
        () => {
            loop {
                match tokens.get(0) {
                    Some((kind, _)) if kind.at_trivia() => {
                        advance!();
                    }
                    _ => break,
                }
            }
        };
    }

    for event in events {
        match event {
            Event::Start(kind) => {
                builder.start_node(kind.into());
                trivia!();
            }
            Event::Token => {
                advance!();
                trivia!();
            }
            Event::End => builder.finish_node(),
            Event::Placeholder => {
                //
            }
        }
    }

    SyntaxNode::new_root(builder.finish())
}
