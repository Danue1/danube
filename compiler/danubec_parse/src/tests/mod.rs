use crate::parse;
use danubec_diagnostic::Diagnostic;

#[test]
fn top_level_attribute() {
    let source = r#"
#![a("value")]
#![a(key)]
#![a(key1, key2)]
#![a(key = "value")]
#![a::b(key = "value")]"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn attribute() {
    let source = r#"
#[a("value")]
struct Foo;"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn zero_sized_struct() {
    let source = r#"
struct Foo;"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn named_struct() {
    let source = r#"
struct Foo { }
struct Bar {
    a: i32,
    b: String,
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn unnamed_struct() {
    let source = r#"
struct Foo();
struct Bar(());
struct Baz(i32, String);"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn mixed_enum() {
    let source = r#"
enum Foo {}
enum Bar {
    A,
    B { a: i32, b: String },
    C(i32, String),
    D = 42,
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn module() {
    let source = r#"
mod foo;
mod bar {
    struct Foo;
    enum Bar {
        A,
        B { a: i32, b: String },
        C(i32, String),
        D = 42,
    }
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn constant() {
    let source = r#"
const FOO: i32 = 42;
const BAR: str = "Hello, world!";"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn static_variable() {
    let source = r#"
static FOO: i32 = 42;
static BAR: str = "Hello, world!";"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn type_alias() {
    let source = r#"
type Foo = i32;
type Bar = (i32, String);"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn function() {
    let source = r#"
fn foo();
fn bar() {}
fn baz(a: i32, b: String) -> (i32, String) {
    (a, b)
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    dbg!(&diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn trait_definition() {
    let source = r#"
trait Foo { }
trait Bar {
    fn foo();
    const BAR: i32;
    type Baz;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn anonymous_implement() {
    let source = r#"
impl Foo { }
impl Foo {
    fn foo() {}
    const BAR: i32 = 42;
    type Baz = i32;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn targeted_implement() {
    let source = r#"
impl Foo for Bar { }
impl Foo for Bar {
    fn foo() {}
    const BAR: i32 = 42;
    type Baz = i32;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn use_definition() {
    let source = r#"
use {};
use {{},*,foo,bar as baz,::{},::*,::foo,::bar as baz,};
use *;
use foo;
use foo as bar;
use foo::{};
use foo::{*,foo,bar as baz,{},::*,::foo,::bar as baz,::{}};

use ::{};
use ::{{},*,foo,bar as baz,::{},::*,::foo,::bar as baz,};
use ::*;
use ::foo;
use ::foo as bar;
use ::foo::{};
use ::foo::{*,foo,bar as baz,{},::*,::foo,::bar as baz,::{}};
"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    dbg!(&diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}
