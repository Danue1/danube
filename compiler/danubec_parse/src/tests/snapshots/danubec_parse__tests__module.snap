---
source: compiler/danubec_parse/src/tests/mod.rs
expression: events
---
KRATE_NODE@0..142
  NEW_LINE@0..1 "\n"
  MODULE_DEFINITION_NODE@1..10
    MOD@1..4 "mod"
    WHITESPACE@4..5 " "
    IDENTIFIER_NODE@5..8
      IDENTIFIER_SEGMENT@5..8
        IDENTIFIER@5..8 "foo"
    MODULE_DEFINITION_OUTLINE_NODE@8..10
      SEMICOLON@8..9 ";"
      NEW_LINE@9..10 "\n"
  MODULE_DEFINITION_NODE@10..142
    MOD@10..13 "mod"
    WHITESPACE@13..14 " "
    IDENTIFIER_NODE@14..18
      IDENTIFIER_SEGMENT@14..18
        IDENTIFIER@14..17 "bar"
        WHITESPACE@17..18 " "
    MODULE_DEFINITION_INLINE_NODE@18..142
      LEFT_BRACE@18..19 "{"
      NEW_LINE@19..20 "\n"
      WHITESPACE@20..24 "    "
      STRUCT_DEFINITION_NODE@24..40
        STRUCT@24..30 "struct"
        WHITESPACE@30..31 " "
        IDENTIFIER_NODE@31..34
          IDENTIFIER_SEGMENT@31..34
            IDENTIFIER@31..34 "Foo"
        STRUCT_BODY_UNIT_NODE@34..40
          SEMICOLON@34..35 ";"
          NEW_LINE@35..36 "\n"
          WHITESPACE@36..40 "    "
      ENUM_DEFINITION_NODE@40..141
        ENUM@40..44 "enum"
        WHITESPACE@44..45 " "
        IDENTIFIER_NODE@45..49
          IDENTIFIER_SEGMENT@45..49
            IDENTIFIER@45..48 "Bar"
            WHITESPACE@48..49 " "
        LEFT_BRACE@49..50 "{"
        NEW_LINE@50..51 "\n"
        WHITESPACE@51..59 "        "
        ENUM_VARIANT_UNIT_NODE@59..60
          IDENTIFIER_NODE@59..60
            IDENTIFIER_SEGMENT@59..60
              IDENTIFIER@59..60 "A"
        COMMA@60..61 ","
        NEW_LINE@61..62 "\n"
        WHITESPACE@62..70 "        "
        ENUM_VARIANT_NAMED_NODE@70..93
          IDENTIFIER_NODE@70..72
            IDENTIFIER_SEGMENT@70..72
              IDENTIFIER@70..71 "B"
              WHITESPACE@71..72 " "
          LEFT_BRACE@72..73 "{"
          WHITESPACE@73..74 " "
          ENUM_VARIANT_NAMED_FIELD_NODE@74..80
            IDENTIFIER_NODE@74..75
              IDENTIFIER_SEGMENT@74..75
                IDENTIFIER@74..75 "a"
            COLON@75..76 ":"
            WHITESPACE@76..77 " "
            PATH_TYPE_NODE@77..80
              PATH_NODE@77..80
                PATH_SEGMENT_IDENTIFIER_NODE@77..80
                  IDENTIFIER_NODE@77..80
                    IDENTIFIER_SEGMENT@77..80
                      IDENTIFIER@77..80 "i32"
          COMMA@80..81 ","
          WHITESPACE@81..82 " "
          ENUM_VARIANT_NAMED_FIELD_NODE@82..92
            IDENTIFIER_NODE@82..83
              IDENTIFIER_SEGMENT@82..83
                IDENTIFIER@82..83 "b"
            COLON@83..84 ":"
            WHITESPACE@84..85 " "
            PATH_TYPE_NODE@85..92
              PATH_NODE@85..92
                PATH_SEGMENT_IDENTIFIER_NODE@85..92
                  IDENTIFIER_NODE@85..92
                    IDENTIFIER_SEGMENT@85..92
                      IDENTIFIER@85..91 "String"
                      WHITESPACE@91..92 " "
          RIGHT_BRACE@92..93 "}"
        COMMA@93..94 ","
        NEW_LINE@94..95 "\n"
        WHITESPACE@95..103 "        "
        ENUM_VARIANT_UNNAMED_NODE@103..117
          IDENTIFIER_NODE@103..104
            IDENTIFIER_SEGMENT@103..104
              IDENTIFIER@103..104 "C"
          LEFT_PAREN@104..105 "("
          PATH_TYPE_NODE@105..108
            PATH_NODE@105..108
              PATH_SEGMENT_IDENTIFIER_NODE@105..108
                IDENTIFIER_NODE@105..108
                  IDENTIFIER_SEGMENT@105..108
                    IDENTIFIER@105..108 "i32"
          COMMA@108..109 ","
          WHITESPACE@109..110 " "
          PATH_TYPE_NODE@110..116
            PATH_NODE@110..116
              PATH_SEGMENT_IDENTIFIER_NODE@110..116
                IDENTIFIER_NODE@110..116
                  IDENTIFIER_SEGMENT@110..116
                    IDENTIFIER@110..116 "String"
          RIGHT_PAREN@116..117 ")"
        COMMA@117..118 ","
        NEW_LINE@118..119 "\n"
        WHITESPACE@119..127 "        "
        ENUM_VARIANT_SCALAR_NODE@127..133
          IDENTIFIER_NODE@127..129
            IDENTIFIER_SEGMENT@127..129
              IDENTIFIER@127..128 "D"
              WHITESPACE@128..129 " "
          EQUAL@129..130 "="
          WHITESPACE@130..131 " "
          LITERAL_EXPRESSION_NODE@131..133
            INTEGER_LITERAL_NODE@131..133
              INTEGER_SEGMENT@131..133 "42"
        COMMA@133..134 ","
        NEW_LINE@134..135 "\n"
        WHITESPACE@135..139 "    "
        RIGHT_BRACE@139..140 "}"
        NEW_LINE@140..141 "\n"
      RIGHT_BRACE@141..142 "}"
